### PR TITLE
[SPARK-44493][SQL] Support for translating catalyst expressions into partial datasource filters

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Since SPARK-22548, Spark disables translating catalyst expressions to partial datasource filters. In fact, sometimes it can be translated as a partial data source filter.

Similar to SPARK-27698, this PR enhances `DataSourceStrategy.translateFilter` to support translate to partial predicates and pushdown to datasource. For example:
```
CREATE TABLE foobar (name string, theid int) using parquet;
EXPLAIN EXTENDED SELECT * FROM foobar WHERE (theid > 0 AND TRIM(name) = 'mary') OR (theid > 10);

== Optimized Logical Plan ==
Filter (((THEID#1 > 0) AND (trim(NAME#0, None) = mary)) OR (NAME#0 = fred))
+- Relation spark_catalog.default.foobar[name#0,theid#1] parquet

== Physical Plan ==
*(1) Filter (((THEID#1 > 0) AND (trim(NAME#0, None) = mary)) OR (NAME#0 = fred))
+- *(1) ColumnarToRow
   +- FileScan parquet spark_catalog.default.foobar[name#0,theid#1] Batched: true, DataFilters: [(((theid#1 > 0) AND (trim(name#0, None) = mary)) OR (name#0 = fred))], Format: Parquet, PartitionFilters: [], PushedFilters: [Or(GreaterThan(theid,0),EqualTo(name,fred))], ReadSchema: struct<name:string,theid:int>
```

But not all datasources support translate to partial predicates. For example, the original filter node is missing if a partial predicate is pushed down to JDBC, which will generate wrong results:
```
EXPLAIN EXTENDED SELECT * FROM foobar WHERE (THEID > 0 AND TRIM(NAME) = 'mary') OR (THEID > 10);

== Optimized Logical Plan ==
Filter (((THEID#1 > 0) AND (trim(NAME#0, None) = mary)) OR (THEID#1 > 10))
+- Relation [NAME#0,THEID#1] JDBCRelation(TEST.PEOPLE) [numPartitions=1]

== Physical Plan ==
*(1) Scan JDBCRelation(TEST.PEOPLE) [numPartitions=1] [NAME#0,THEID#1] PushedFilters: [*Or(GreaterThan(THEID,0),GreaterThan(THEID,10))], ReadSchema: struct<NAME:string,THEID:int>
```

Supported translate to partial predicates:
https://github.com/apache/spark/blob/679ea56dc8be7566c32a606639aa052421136afc/sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala#L383
https://github.com/apache/spark/blob/2994560d204c70cd17424e130dad5e3ece0988f5/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala#L202
https://github.com/apache/spark/blob/c4cd193b4d43d951255d9914f9b6f099d140fe1c/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScanBuilder.scala#L80

Unsupported translate to partial predicates:
https://github.com/apache/spark/blob/8ff6b7a04cbaef9c552789ad5550ceab760cb078/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala#L674
https://github.com/apache/spark/blob/6161bf44f40f8146ea4c115c788fd4eaeb128769/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownUtils.scala#L55-L56

### Why are the changes needed?

Pushdown more filters to improve query performance.

### Does this PR introduce _any_ user-facing 

No.

### How was this patch tested?

Unit test and benchmark result:

Before | After
-- | --
image | image
image | image
image | image

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan

**Comment by wangyum** [comment]
@MaxGekk @huaxingao @gengliangwang @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 1]
```diff
@@ -573,8 +573,10 @@ object DataSourceStrategy
    * @return a `Some[Filter]` if the input [[Expression]] is convertible, otherwise a `None`.
    */
   protected[sql] def translateFilter(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:575]: can we rename it? It's not pure translation anymore, but kind of extracting filters to do pushdown

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:575]: do we still use it for pure translation?

Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:575]: Are these two pure translations?

https://github.com/apache/spark/blob/071feabbd4325504332679dfa620bc5ee4359370/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala#L674
https://github.com/apache/spark/blob/6161bf44f40f8146ea4c115c788fd4eaeb128769/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownUtils.scala#L54-L56

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 2]
```diff
@@ -584,46 +586,46 @@ object DataSourceStrategy
...
+   * @param supportNestedPushDown Whether nested predicate push down is enabled.
+   * @param canPartialPushDown Can it be translated into partial predicate.
    * @return a `Some[Filter]` if the input [[Expression]] is convertible, otherwise a `None`.
    */
   protected[sql] def translateFilterWithMapping(
       predicate: Expression,
       translatedFilterToExpr: Option[mutable.HashMap[sources.Filter, Expression]],
-      nestedPredicatePushdownEnabled: Boolean)
+      supportNestedPushDown: Boolean,
+      canPartialPushDown: Boolean)
```
Comment (melihsozdinler) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:597]: You can define default value as false, and send the parameter on demand if true is required by this function.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 3]
```diff
@@ -573,8 +573,10 @@ object DataSourceStrategy
    * @return a `Some[Filter]` if the input [[Expression]] is convertible, otherwise a `None`.
    */
   protected[sql] def translateFilter(
-      predicate: Expression, supportNestedPredicatePushdown: Boolean): Option[Filter] = {
-    translateFilterWithMapping(predicate, None, supportNestedPredicatePushdown)
+      predicate: Expression,
+      supportNestedPushDown: Boolean,
+      canPartialPushDown: Boolean): Option[Filter] = {
```
Comment (melihsozdinler) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:578]: canPartialPushdown and supportNestedPushDown can have a default value as false, so if some logic does not require these options, no need to define variable before calling function.

**Pull request closed without merging**